File: Ruby/lib/reverse_integer.rb
# frozen_string_literal: true

# @param {Integer} x
# @return {Integer}
def reverse(x)
  result = 0
  is_negative = x.negative?
  x = x.abs if is_negative

  while x != 0
    last_number = x % 10
    result = result * 10 + last_number

    return 0 if result > 2**31 - 1 || result < -2**31

    x = (x / 10).floor
  end

  is_negative ? -result : result
end

File: JavaScript/src/reverse_integer.js
/**
 * @param {number} x
 * @return {number}
 */
function reverse(x) {
  let result = 0;
  const MAX_INT32 = 2 ** 31 - 1;
  const MIN_INT32 = -MAX_INT32 - 1;
  const isNegative = x < 0;
  let temp = x;

  if (isNegative) {
    if (temp === MIN_INT32) {
      return 0;
    }
    temp = -temp;
  }

  while (temp !== 0) {
    const lastDigit = temp % 10;
    result = result * 10 + lastDigit;
    if (result > MAX_INT32 || result < MIN_INT32) {
      return 0;
    }
    temp = Math.floor(temp / 10);
  }

  return isNegative ? -result : result;
}

export { reverse };
